File: PODM/SW/aaa-service/src/main/java/com/intel/rsd/podm/services/SessionService.java
/*
 * Copyright (c) 2019 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.rsd.podm.services;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.intel.rsd.podm.security.RedfishSessionCreator;
import com.intel.rsd.podm.sessions.RedfishSession;
import com.intel.rsd.podm.sessions.RedfishSessionsRepository;
import com.intel.rsd.podm.sessions.SessionResource;
import lombok.val;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.annotation.PostConstruct;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

import static com.intel.rsd.json.JsonUtils.readJsonFromFile;
import static com.intel.rsd.podm.sessions.RedfishSession.X_AUTH_TOKEN;
import static com.intel.rsd.redfish.RedfishUtils.toODataIdArrayNode;
import static java.util.stream.Collectors.toList;
import static org.springframework.http.HttpHeaders.LOCATION;
import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.http.HttpStatus.OK;
import static org.springframework.http.ResponseEntity.noContent;

@RestController
@SuppressWarnings({"checkstyle:ClassFanOutComplexity"})
@RequestMapping("/redfish/v1")
public class SessionService {

    @Value("${aaa-config.session-timeout}")
    private Long sessionTimeout;

    private ObjectNode sessionServiceResource;
    private ObjectNode sessionsCollectionResource;

    private RedfishSessionsRepository sessionsRepository;
    private RedfishSessionCreator sessionCreator;

    @Autowired
    public SessionService(RedfishSessionsRepository sessionsRepository, RedfishSessionCreator tokenGranter) {
        this.sessionsRepository = sessionsRepository;
        this.sessionCreator = tokenGranter;
    }

    @PostConstruct
    public void postConstruct() throws IOException {
        ObjectNode sessionRootTemplate = readJsonFromFile(this, "SessionServiceTemplate.json");
        sessionServiceResource = sessionRootTemplate.put("SessionTimeout", sessionTimeout);
        sessionsCollectionResource = readJsonFromFile(this, "SessionsCollectionTemplate.json");
    }

    @GetMapping("SessionService")
    public ObjectNode getSessionService() {
        return sessionServiceResource;
    }

    @PostMapping("SessionService/Sessions")
    public ResponseEntity createSession(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse) {
        RedfishSession redfishSession = sessionCreator.createSession(httpServletRequest);
        if (redfishSession == null) {
            throw new SessionCreationException("Unable to create session for provided credentials.");
        }
        httpServletResponse.setHeader(X_AUTH_TOKEN, redfishSession.getXAuthToken());
        httpServletResponse.setHeader(LOCATION, "/redfish/v1/SessionService/Sessions/" + redfishSession.getId());
        return ResponseEntity.status(CREATED).body(new SessionResource(redfishSession.getId(), redfishSession.getUserName()));
    }

    @GetMapping("SessionService/Sessions")
    public JsonNode getSessions() {
        val ids = sessionsRepository.getSessionIds().stream().map(id -> "/redfish/v1/SessionService/Sessions/" + id).collect(toList());
        val idsColl = toODataIdArrayNode(ids);
        return sessionsCollectionResource.deepCopy()
            .put("Members@odata.count", idsColl.size())
            .set("Members", idsColl);
    }

    @GetMapping("SessionService/Sessions/{id}")
    public ResponseEntity getSession(@PathVariable String id) {
        RedfishSession redfishSession = sessionsRepository.findById(id);
        if (redfishSession == null) {
            throw new ResourceNotFoundException("Unable to find session for given id.");
        }

        SessionResource session = new SessionResource(id, redfishSession.getUserName());
        return new ResponseEntity<>(session, OK);
    }

    @DeleteMapping("SessionService/Sessions/{id}")
    public ResponseEntity deleteSession(@PathVariable("id") String id) {
        boolean sessionRemoved = sessionsRepository.removeSession(id);
        if (sessionRemoved) {
            return noContent().build();
        }
        throw new ResourceNotFoundException("Unable to find session for given id.");
    }
}

File: PSME/application/src/rest/security/session/session_service.cpp
/*!
 * @copyright
 * Copyright (c) 2018-2019 Intel Corporation
 *
 * @copyright
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * @copyright
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * @copyright
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file session_service.cpp
 *
 * @brief SessionService implementation
 *
 * */

#include "psme/rest/security/session/session_service.hpp"
#include "psme/rest/security/session/session_manager.hpp"
#include "logger/logger.hpp"



using namespace psme::rest::security::session;
using std::chrono::steady_clock;


void SessionService::start() {
    log_info("rest", "Starting REST session service ...");
    if (!is_running()) {
        Thread::start();
        log_info("rest", "REST session service started.");
    }
}


void SessionService::stop() {
    log_info("rest", "Stopping REST session service ...");
    if (is_running()) {
        Thread::stop();
        log_info("rest", "REST session service stopped.");
    }
}


SessionService::~SessionService() {
    stop();
}


void SessionService::handle_sessions() {
    while (is_running()) {

        auto start_point = steady_clock::now();

        try {
            SessionManager::get_instance()->remove_outdated_sessions();
        }
        catch (const std::runtime_error& e) {
            log_error("rest",
                      " Exception occurred when cleaning the session container: " << e.what());
        }
        catch (...) {
            log_error("rest",
                      " Exception occurred when cleaning the session container.");
        }
        auto sleep_period = m_session_clenup_interval - (steady_clock::now() - start_point);

        std::this_thread::sleep_for(sleep_period);
    }
}
